File: Leet-611. Valid Triangle Number/solution-n^2.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var triangleNumber = function(nums) {
    if (nums === null || nums.length < 3) {
        return 0;
    }

    nums.sort(function(a, b) {
        return a - b;
    });
    
    var res = 0;

    for (var i = nums.length - 1; i >= 2; i--) {
        var left = 0;
        var right = i - 1;
        while (left < right) {
            if (nums[left] + nums[right] > nums[i]) {
                res += right - left;
                right--;
            } else {
               left++; 
            }
        }
    }

    return res;
};

File: Leet-611. Valid Triangle Number/solution-n^2.java
class Solution {
    public int triangleNumber(int[] nums) {
        if (nums == null || nums.length < 3) {
            return 0;
        }
        
        Arrays.sort(nums);
        int res = 0;
        
        for (int i = nums.length - 1; i >= 2; i--) {
            int left = 0, right = i - 1;
            while (left < right) {
                if (nums[left] + nums[right] > nums[i]) {
                    res += right - left;
                    right--;
                } else {
                    left++;
                }
            }
        }
        
        return res;
    }
}